feat: management command to purge information about web certificates

deborahgu opened the pull request:
for Open edX  operators who still have users with legacy PDF certificates, retirement requires first extracting information from the user's GeneratedCertificate record in order to delete the 4 associated files for each PDF certificate, and then removing the links to the relevant files.  this creates a management command to do the second half of that work, to be called after the file cleanup has been completed.

FIXES: APER-3889

<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Describe what this pull request changes, and why. Include implications for people using this change.
Design decisions and their rationales should be documented in the repo (docstring / ADR), per
[OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/oep-0019-bp-developer-documentation.html), and can be
linked here.

Useful information to include:

- Which edX user roles will this change impact? Common user roles are "Learner", "Course Author",
"Developer", and "Operator".
- Include screenshots for changes to the UI (ideally, both "before" and "after" screenshots, if applicable).
- Provide links to the description of corresponding configuration changes. Remember to correctly annotate these
changes.

## Supporting information

Link to other information about the change, such as Jira issues, GitHub issues, or Discourse discussions.
Be sure to check they are publicly readable, or if not, repeat the information here.

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.

- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

deborahgu committed:
feat: management command to purge information about web certificates from retired users

for Open edX  operators who still have users with legacy PDF
certificates, retirement requires first extracting information from
the user's GeneratedCertificate record in order to delete the 4
associated files for each PDF certificate, and then removing the links
to the relevant files.  this creates a management command to do that
work.

FIXES: APER-3889

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3889_management-command-to-nullify-download_url-in-a-cert

deborahgu committed:
feat: management command

removing unused import

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3889_management-command-to-nullify-download_url-in-a-cert

justinhynes commented on lms/djangoapps/certificates/management/commands/purge_references_to_pdf_certificates.py:
```diff
@@ -0,0 +1,106 @@
+    Invokes the custom save() function defined as part of the `GeneratedCertificate`
+    model; perforce will emit no Django signals. This is desired behavior. We are
+    using this management command to purge information that was never sent to any
+    other systems, so we don't need to propagate updates.
```
I'm not sure if I'm misunderstanding this comment.

My understanding is that when you use a filter to update multiple records it will bypass Django's model methods and _won't_ call our custom/overridden save() function.

In this case, wouldn't we want to? We're updating the status of the certificate records to `unavailable` which could create a mismatch between the LMS and credentials. If we weren't updating the status, then I don't think this is as big of a deal. 

Would we instead want to loop through each certificate record and call the `invalidate()` function (https://github.com/openedx/edx-platform/blob/6c6fd84e533e6cb272da2e618f227225ad907232/lms/djangoapps/certificates/models.py#L310-L325)? 

This would allow the system to call the `_revoke_certificate()` method and wipe out the `download_uuid`, `download_url`, and update the records status in additional to emitting the event that would revoke the record in Credentials.

deborahgu replied:
oh yeah, I decided to update the status after I decided on the algorithm. will update.

justinhynes replied:
After reading your most recent commit's message, I agree with what you're saying. This does help us get achieve better data integrity in an imperfect system where maintaining support for legacy PDF certificates earned by learners has been a bit... messy. 👍

deborahgu committed:
feat: updating management command

After thinking about it, I have removed the update to `status` from this
management command, as per the original specification of the ticket. I
added it for completeness originally, but was already uncomfortable,
because it's not exactly accurate. The `CertificateStatuses` enum does
define a `deleted` status:

```
    deleted             - The PDF certificate has been deleted.
```

but I think it's inappropriate to use here.

There are multiple places in the code where it's clear that almost all of the statuses are legacy and unused (eg. [Example 1](https://github.com/openedx/edx-platform/blob/6c6fd84e533e6cb272da2e618f227225ad907232/lms/djangoapps/certificates/data.py#L12-L34), [Example 2](https://github.com/openedx/edx-platform/blob/1029de5537752ecf0477d578dba8a932e741e497/common/djangoapps/student/helpers.py#L491-L492)). There are innumerable APIs in the system that have expectations about what might possibly be returned from a `GeneratedCertificate.status` object, and none of them is expecting `deleted`

Ultimately, the certificate isn't revoked, which has a specific meaning around saying it was unearned. The certificate was earned; it has simply been deleted.  We should not be kicking off program certificate invalidation, because that's not what's happening. We should be trusting the normal user retirement process to remove/purge PII from any program certificates that might exist. The nature of web certificates simply means that we are going through this process outside of the normal retirement flow. The normal retirement flow can be trusted to implement any certificate object revocation/removal/PII-purging, and doing an extra step outside of that flow is counterproductive.

When PDF certificates were removed from the system, they weren't removed in their entirety. Instead, we have this vestigial remains of PDF certificates code, just enough to allow learners to display and use the ones that they already have, without any of the support systems for modifying them. Adding a `deleted` status, verifying that all other APIs wouldn't break in the presence of a certificate with that status, adding the signals to process and propagate the change: all of this would be adding more tech debt upon the already existing technical debt which is the PDF certs code. Better to simply add this one necessary data integrity change, and focus on a process which might allow us to eventually remove the web certificates code.

The original ask was simply to enforce data integrity: to remove links to files that have been deleted, as an indication that they've been deleted. I only added `status` update out of a (misplaced but well-intentioned) completionist urge.

deborahgu committed:
feat:  management command

fixing a bad comment
